Sources/StreamChatUI/ChatThread/ChatThreadVC.swift

```diff
@@ -0,0 +1,227 @@
+         and non-evaluated elements.
+
+         We have evaluated thread root message in `messageController.message` but to get keep lazy map
+         working after an insert we also need an underlaying DTO to be added to source collection and it's getting
+         hard since the information about source collection `Element` type is available only during lazy map
+         initialization and does not get stored for later use.
+
+         It could be addressed on LLC side by tweaking an observer to fetch thread root message along with replies.
+         */
+        let replies = Array(messageController.replies)
+
+        if let threadRootMessage = messageController.message {
+            return replies + [threadRootMessage]
+        }
+
+        return replies
+    }
+
+    open func channel(for vc: ChatMessageListVC) -> ChatChannel? {
+        channelController.channel
+    }
+
+    open func numberOfMessages(in vc: ChatMessageListVC) -> Int {
+        replies.count
+    }
+
+    open func chatMessageListVC(_ vc: ChatMessageListVC, messageAt indexPath: IndexPath) -> ChatMessage? {
+        guard indexPath.item < replies.count else { return nil }
+        return replies[indexPath.item]
+    }
+
+    open func chatMessageListVC(
+        _ vc: ChatMessageListVC,
+        messageLayoutOptionsAt indexPath: IndexPath
+    ) -> ChatMessageLayoutOptions {
+        guard let channel = channelController.channel else { return [] }
+
+        var layoutOptions = components
+            .messageLayoutOptionsResolver
+            .optionsForMessage(
+                at: indexPath,
+                in: channel,
+                with: AnyRandomAccessCollection(replies),
+                appearance: appearance
+            )
+
+        layoutOptions.remove(.threadInfo)
+
+        return layoutOptions
+    }
+}
+
+extension ChatThreadVC: ChatMessageListVCDelegate {
+    open func chatMessageListVC(
+        _ vc: ChatMessageListVC,
+        willDisplayMessageAt indexPath: IndexPath
+    ) {
+        if messageController.state == .remoteDataFetched && indexPath.row == replies.count - 5 {
+            messageController.loadPreviousReplies()
+        }
+    }
+
+    open func chatMessageListVC(
+        _ vc: ChatMessageListVC,
+        didTapOnAction actionItem: ChatMessageActionItem,
+        for message: ChatMessage
+    ) {
+        switch actionItem {
+        case is EditActionItem:
+            dismiss(animated: true) { [weak self] in
+                self?.messageComposerVC.content.editMessage(message)
+            }
+        case is InlineReplyActionItem:
+            dismiss(animated: true) { [weak self] in
+                self?.messageComposerVC.content.quoteMessage(message)
+            }
+        default:
+            return
+        }
+    }
+
+    open func chatMessageListVC(
+        _ vc: ChatMessageListVC,
+        scrollViewDidScroll scrollView: UIScrollView
+    ) {
+        // No-op
```

Comment:
do we have a marking policy for such comments? I would assume `Note: Do nothing cause its a UI requirement` will looks better, or feel free to propose something in this approach 

>https://github.com/raywenderlich/swift-style-guide#comments

Reply:
Will prove a better comment 👍

Reply:
✅